In [3]:
#pip install duckdb

     --------------------------------------- 10.3/10.3 MB 21.1 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd

credit_data = pd.read_csv("credit_mergedata_new.csv")
credit_data.head()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0  5008827      M       Y            Y               0       180000.0   
1  5009744      F       Y            N               0       315000.0   
2  5009746      F       Y            N               0       315000.0   
3  5009749      F       Y            N               0       166500.0   
4  5009752      F       Y            N               0       315000.0   

      Employment_status         Education Marital_status       Housing_type  \
0             Pensioner  Higher education        Married  House / apartment   
1  Commercial associate  Higher education        Married  House / apartment   
2  Commercial associate  Higher education        Married  House / apartment   
3  Commercial associate  Higher education        Married  House / apartment   
4  Commercial associate  Higher education        Married  House / apartment   

   Has_mobile  Has_work_phone  Has_phone  Has_email_id Type_Occupation  \
0           1               0          0             0             NaN   
1           1               1          1             0             NaN   
2           1               1          1             0             NaN   
3           1               1          1             0             NaN   
4           1               1          1             0             NaN   

   Family_Members  Is_approved    Age  Employed_years  
0               2            0  51.00             0.0  
1               2            0  37.00             1.0  
2               2            0  43.42             1.0  
3               2            0  37.00             1.0  
4               2            0  37.00             1.0

In [3]:
import duckdb

conn = duckdb.connect()
conn.register('credit_data', credit_data)


In [4]:
conn.execute("select * from credit_data").fetchdf().head()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0  5008827      M       Y            Y               0       180000.0   
1  5009744      F       Y            N               0       315000.0   
2  5009746      F       Y            N               0       315000.0   
3  5009749      F       Y            N               0       166500.0   
4  5009752      F       Y            N               0       315000.0   

      Employment_status         Education Marital_status       Housing_type  \
0             Pensioner  Higher education        Married  House / apartment   
1  Commercial associate  Higher education        Married  House / apartment   
2  Commercial associate  Higher education        Married  House / apartment   
3  Commercial associate  Higher education        Married  House / apartment   
4  Commercial associate  Higher education        Married  House / apartment   

   Has_mobile  Has_work_phone  Has_phone  Has_email_id Type_Occupation  \
0           1               0          0             0            None   
1           1               1          1             0            None   
2           1               1          1             0            None   
3           1               1          1             0            None   
4           1               1          1             0            None   

   Family_Members  Is_approved    Age  Employed_years  
0               2            0  51.00             0.0  
1               2            0  37.00             1.0  
2               2            0  43.42             1.0  
3               2            0  37.00             1.0  
4               2            0  37.00             1.0

# 1. Group the customers based on their income type and find the average of        their annual income.

In [8]:
conn.execute("SELECT Employment_status, ROUND(AVG(Annual_income), 2) AS average_annual_income FROM credit_data GROUP BY Employment_status ORDER BY average_annual_income DESC").fetchdf()

Employment_status  average_annual_income
0  Commercial associate              233107.40
1         State servant              211422.41
2               Working              180848.21
3             Pensioner              155343.50

# 2. Find the female owners of cars and property.

In [9]:
conn.execute("SELECT * FROM credit_data WHERE Gender = 'F' AND Has_car = 'Y' AND Has_property = 'Y'").fetchdf()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0    5018498      F       Y            Y               0        90000.0   
1    5018501      F       Y            Y               0       166500.0   
2    5018503      F       Y            Y               0        90000.0   
3    5024213      F       Y            Y               0       540000.0   
4    5036660      F       Y            Y               0        76500.0   
..       ...    ...     ...          ...             ...            ...   
174  5048458      F       Y            Y               1       126000.0   
175  5023719      F       Y            Y               0       175500.0   
176  5033520      F       Y            Y               3       180000.0   
177  5024049      F       Y            Y               1       144000.0   
178  5053790      F       Y            Y               0       225000.0   

        Employment_status                      Education Marital_status  \
0                 Working  Secondary / secondary special        Married   
1                 Working  Secondary / secondary special        Married   
2                 Working  Secondary / secondary special        Married   
3    Commercial associate               Higher education        Married   
4               Pensioner  Secondary / secondary special        Married   
..                    ...                            ...            ...   
174               Working               Higher education        Married   
175             Pensioner               Higher education        Married   
176               Working  Secondary / secondary special        Married   
177               Working               Higher education        Married   
178               Working               Higher education        Married   

            Housing_type  Has_mobile  Has_work_phone  Has_phone  Has_email_id  \
0      House / apartment           1               1          1             0   
1      House / apartment           1               1          1             0   
2      House / apartment           1               1          1             0   
3      House / apartment           1               0          1             0   
4      House / apartment           1               0          0             0   
..                   ...         ...             ...        ...           ...   
174    House / apartment           1               0          0             0   
175    House / apartment           1               0          1             0   
176  Municipal apartment           1               0          1             0   
177    House / apartment           1               0          0             0   
178    House / apartment           1               0          0             0   

    Type_Occupation  Family_Members  Is_approved   Age  Employed_years  
0     Cooking staff               2            0  51.0             2.0  
1     Cooking staff               2            0  51.0             2.0  
2     Cooking staff               2            0  51.0             2.0  
3              None               2            0  42.0             0.0  
4              None               2            0  65.0             0.0  
..              ...             ...          ...   ...             ...  
174            None               3            1  25.0             0.0  
175            None               2            1  60.0             0.0  
176  Medicine staff               5            1  34.0            16.0  
177     Accountants               3            1  35.0             8.0  
178            None               2            1  45.0             7.0  

[179 rows x 19 columns]

# 3. Find the male customers who are staying with their families.

In [10]:
conn.execute("SELECT * FROM credit_data WHERE Gender ='M' AND Family_Members > 1").fetchdf()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0    5008827      M       Y            Y               0       180000.0   
1    5010864      M       Y            Y               1       450000.0   
2    5010868      M       Y            Y               1       450000.0   
3    5021303      M       N            N               1       472500.0   
4    5021310      M       N            Y               0       270000.0   
..       ...    ...     ...          ...             ...            ...   
465  5096856      M       Y            Y               0       180000.0   
466  5090942      M       N            N               0       225000.0   
467  5118268      M       Y            N               1       360000.0   
468  5115992      M       Y            Y               2       180000.0   
469  5118219      M       Y            N               0       270000.0   

        Employment_status                      Education  Marital_status  \
0               Pensioner               Higher education         Married   
1    Commercial associate  Secondary / secondary special         Married   
2               Pensioner  Secondary / secondary special         Married   
3               Pensioner               Higher education         Married   
4                 Working  Secondary / secondary special         Married   
..                    ...                            ...             ...   
465  Commercial associate  Secondary / secondary special         Married   
466  Commercial associate  Secondary / secondary special         Married   
467         State servant  Secondary / secondary special         Married   
468               Working               Higher education         Married   
469               Working  Secondary / secondary special  Civil marriage   

          Housing_type  Has_mobile  Has_work_phone  Has_phone  Has_email_id  \
0    House / apartment           1               0          0             0   
1    House / apartment           1               0          1             1   
2    House / apartment           1               0          1             1   
3         With parents           1               0          0             1   
4    House / apartment           1               0          0             0   
..                 ...         ...             ...        ...           ...   
465  House / apartment           1               0          0             0   
466  House / apartment           1               0          0             0   
467  House / apartment           1               0          1             0   
468  House / apartment           1               0          0             0   
469  House / apartment           1               1          1             0   

        Type_Occupation  Family_Members  Is_approved   Age  Employed_years  
0                  None               2            0  51.0             0.0  
1            Core staff               3            0  49.0             1.0  
2            Core staff               3            0  49.0             1.0  
3                  None               3            0  24.0             2.0  
4              Laborers               2            0  46.0             0.0  
..                  ...             ...          ...   ...             ...  
465             Drivers               2            1  49.0             9.0  
466  Low-skill Laborers               2            1  36.0             3.0  
467             Drivers               3            1  30.0             9.0  
468            Managers               4            1  36.0             6.0  
469             Drivers               2            1  41.0             1.0  

[470 rows x 19 columns]

In [11]:
conn.execute("SELECT * FROM credit_data WHERE Gender ='M' AND Housing_type = 'With parents'").fetchdf()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0   5021303      M       N            N               1       472500.0   
1   5079166      M       Y            N               0       202500.0   
2   5079167      M       Y            N               0       202500.0   
3   5079168      M       Y            N               0       202500.0   
4   5050729      M       Y            N               0       180000.0   
5   5028383      M       Y            N               0       337500.0   
6   5143019      M       N            N               0       270000.0   
7   5067982      M       Y            Y               0       153000.0   
8   5143573      M       Y            N               0       157500.0   
9   5038751      M       Y            Y               2       157500.0   
10  5010203      M       Y            Y               0       135000.0   
11  5058267      M       N            N               1       225000.0   
12  5024352      M       Y            Y               0       585000.0   
13  5150038      M       N            N               0       180000.0   
14  5113302      M       N            Y               1       180000.0   
15  5126311      M       N            N               0       180000.0   
16  5126751      M       Y            N               2       135000.0   
17  5037125      M       Y            N               0       180000.0   
18  5054414      M       N            N               0       225000.0   
19  5145694      M       Y            Y               0       225000.0   
20  5146429      M       Y            N               0       261000.0   
21  5054407      M       N            N               0       225000.0   
22  5068648      M       N            Y               0       135000.0   
23  5143560      M       N            N               2       202500.0   
24  5125542      M       N            N               0       180000.0   
25  5142781      M       Y            N               2       157500.0   
26  5024592      M       Y            Y               0       157500.0   
27  5094884      M       N            Y               0       450000.0   
28  5033496      M       Y            N               1       202500.0   
29  5067203      M       N            N               1       405000.0   
30  5045537      M       Y            N               2       225000.0   
31  5009269      M       Y            N               0       337500.0   
32  5069007      M       Y            Y               0       337500.0   
33  5066915      M       Y            Y               0       180000.0   
34  5143025      M       N            N               0       270000.0   
35  5010202      M       Y            Y               0       135000.0   
36  5028612      M       Y            Y               0       225000.0   

       Employment_status                      Education        Marital_status  \
0              Pensioner               Higher education               Married   
1                Working  Secondary / secondary special               Married   
2                Working  Secondary / secondary special               Married   
3                Working  Secondary / secondary special               Married   
4                Working  Secondary / secondary special  Single / not married   
5                Working  Secondary / secondary special  Single / not married   
6                Working  Secondary / secondary special               Married   
7                Working               Higher education             Separated   
8                Working              Incomplete higher               Married   
9   Commercial associate               Higher education               Married   
10               Working                Lower secondary               Married   
11  Commercial associate  Secondary / secondary special               Married   
12  Commercial associate  Secondary / secondary special               Married   
13               Working  Secondary / secondary special  S

# 4. Please list the top five people having the highest income.

Note: I have taken the cleaned dataset (after handling null values, renaming columns, extracting columns from existing columns) but not after outlier treatment.

In [13]:
conn.execute("SELECT * FROM credit_data ORDER BY Annual_income DESC LIMIT 5").fetchdf()

Ind_ID Gender Has_car Has_property  Children_count  Annual_income  \
0  5143231      F       Y            Y               1      1575000.0   
1  5143235      F       Y            Y               1      1575000.0   
2  5090470      M       N            Y               1       900000.0   
3  5079016      M       Y            Y               2       900000.0   
4  5079017      M       Y            Y               2       900000.0   

      Employment_status                      Education        Marital_status  \
0  Commercial associate               Higher education  Single / not married   
1  Commercial associate               Higher education  Single / not married   
2               Working  Secondary / secondary special               Married   
3  Commercial associate               Higher education               Married   
4  Commercial associate               Higher education               Married   

        Housing_type  Has_mobile  Has_work_phone  Has_phone  Has_email_id  \
0  House / apartment           1               0          0             0   
1  House / apartment           1               0          0             0   
2  House / apartment           1               0          0             0   
3  House / apartment           1               0          0             0   
4  House / apartment           1               0          0             0   

  Type_Occupation  Family_Members  Is_approved   Age  Employed_years  
0        Managers               2            1  27.0             6.0  
1        Managers               2            1  27.0             6.0  
2        Laborers               3            1  42.0            12.0  
3        Managers               4            1  27.0             2.0  
4        Managers               4            1  27.0             2.0

# 5. How many married people are having bad credit?

In [14]:
conn.execute("SELECT  COUNT(1) AS married_bad_credit_cnt FROM credit_data WHERE Marital_status IN ('Married', 'Civil marriage') AND Is_approved = 0").fetchdf()

married_bad_credit_cnt
0                     118

# 6. What is the highest education level and what is the total count?

In [15]:
conn.execute("SELECT Education, COUNT(1) AS people_cnt FROM credit_data GROUP BY Education").fetchdf()

Education  people_cnt
0              Incomplete higher          68
1                Academic degree           2
2               Higher education         426
3  Secondary / secondary special        1031
4                Lower secondary          21

Highest Education level is Academic Degree with a count of 2 followed by Higher Education with a count of 426.

In [18]:
conn.execute("SELECT COUNT(1) AS people_cnt FROM credit_data WHERE Education = 'Academic degree'").fetchdf()

people_cnt
0           2

# 7. Between married males and females, who is having more bad credit? 

In [19]:
conn.execute("SELECT Gender, COUNT(1) AS gender_bad_credit_cnt FROM credit_data WHERE Is_approved = 0 AND Marital_status IN ('Married', 'Civil marriage') GROUP BY Gender").fetchdf()

Gender  gender_bad_credit_cnt
0      M                     54
1      F                     64

Married males are having more bad credit